src/unity/python/turicreate/toolkits/_mxnet_utils.py

```diff
@@ -99,6 +120,65 @@ def assert_valid_num_gpus():
     _numeric_param_check_range('num_gpus', num_gpus, -1, _six.MAXSIZE)
 
 
+def _warn_if_less_than_cuda_free_memory(mem_mb, max_devices=None, has_batch_size_setting=True):
+    # Note, we cannot use _CUDA_GPUS, since we need an up-to-date snapshot of
+    # the free memory
+    import numpy as np
+    from turicreate.util import _get_cuda_gpus
+    import time
+    import textwrap
+    gpus = _get_cuda_gpus()
+    num_gpus = get_num_gpus_in_use(max_devices=max_devices)
+    gpus = gpus[:num_gpus]
+    capable = all(gpu['memory_total'] > mem_mb for gpu in gpus)
+    available = all(gpu['memory_free'] > mem_mb for gpu in gpus)
+    warning = None
+    if capable and not available:
+        # Get GPU with the least amount of free memroy
+        gpu_mems = [gpu['memory_free'] for gpu in gpus]
+        gpu_idx = np.argmin(gpu_mems)
+
+        warning = '''
+        You may not have enough free GPU memory to create this model, which may
+        result in a fatal crash or lower performance. We recommend that you
+        have at least {mem_req} MiB available on each GPU, while currently you
+        have only {mem_free:.0f} MiB free (out of {mem_total:.0f} MiB) on GPU
+        #{index:d}. Please close all other processes that may be using this
+        GPU.
+        '''.format(mem_req=mem_mb,
+                   mem_free=gpus[gpu_idx]['memory_free'],
+                   mem_total=gpus[gpu_idx]['memory_total'],
+                   index=gpus[gpu_idx]['index'])
+
+    elif not capable:
+        # Get GPU with the least amount of total memroy
+        gpu_mems = [gpu['memory_total'] for gpu in gpus]
+        gpu_idx = np.argmin(gpu_mems)
+
+        warning = '''
+        You may not have a capable enough GPU to create this model with the
+        current settings, which may result in a fatal crash or lower
+        performance. We recommend that you have at least {mem_req} MiB on each
+        requested GPU, while currently GPU #{index} has only {mem_total:.0f}
+        MiB of total memory.
+        '''.format(mem_req=mem_mb,
+                   mem_total=gpus[gpu_idx]['memory_total'],
+                   index=gpus[gpu_idx]['index'])
+
+        if has_batch_size_setting:
+            warning += 'Please try lowering the batch size by passing the parameter batch_size to create().'
+
+    if warning:
+        print('WARNING:')
+        print()
+        # Re-wrap text
+        lines = textwrap.wrap(textwrap.dedent(warning).lstrip(), width=75)
+        for line in lines:
+            print('   ', line)
+        print()
+        time.sleep(3)
```

Comment:
Why the sleep?

Reply:
The thought is that fatal memory crashes typically fill up the screen with a backtrace, so waiting a few seconds will increase the chance of the user seeing this warning. I can remove it if you do not think it will have this effect.

Reply:
Sleeping for 3 seconds is ok. I would just add a comment explaining why.

Reply:
Good idea! Pushed.